pkg/kademlia/kademlia_test.go

```diff
@@ -76,7 +80,8 @@ func TestNewKademlia(t *testing.T) {
 		assert.NoError(t, err)
 		identity, err := ca.NewIdentity()
 		assert.NoError(t, err)
-		actual, err := NewKademlia(v.id, v.bn, v.addr, identity, "db", kc)
+		actual, err := NewKademlia(v.id, v.bn, v.addr, identity, dir, kc)
+		defer cleanup(ctx, t, actual, dir)
```

Comment:
This defer runs at the end of the function so you end up closing while one kademlia instance is still using the folder.

In principle it's doing 
```
kad1.Close(); err := os.RemoveAll("/tmp/kad_test")
kad2.Close(); err := os.RemoveAll("/tmp/kad_test")
```

The first `os.RemoveAll` will fail because it's still in use by `kad2`.

Rather do `defer os.RemoveAll` at the top and don't do the folder cleanup inside.

Similarly the `setup` will fail, because it's being held by the previous Kademlia instance. But, you can remove the setup func by ensuring that each instance will use a different folder like:

```
for i, v := range cases {
    ...
    actual, err := NewKademlia(v.id, v.bn, v.addr, identity, filepath.Join(dir, strconv.Itoa(i)), kc)
```

Reply:
Could you explain what you mean by `setup` further?

Reply:
https://play.golang.org/p/q5UQMmhCNLo <- notice how it never prints "closed". This is because the defers get resolved at the end of the function. This is similar how the os.RemoveAll isn't able to remove files on windows because the files would be still in use (since the shutdown hasn't been run).